// lucid/dist/components/SingleSelect/SingleSelect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.SingleSelectDumb = undefined;

var _isNil2 = require('lodash/isNil');

var _isNil3 = _interopRequireDefault(_isNil2);

var _assign2 = require('lodash/assign');

var _assign3 = _interopRequireDefault(_assign2);

var _isNumber2 = require('lodash/isNumber');

var _isNumber3 = _interopRequireDefault(_isNumber2);

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _first2 = require('lodash/first');

var _first3 = _interopRequireDefault(_first2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

var _stateManagement = require('../../util/state-management');

var _SingleSelect = require('./SingleSelect.reducers');

var reducers = _interopRequireWildcard(_SingleSelect);

var _DropMenu = require('../DropMenu/DropMenu');

var _CaretIcon = require('../Icon/CaretIcon/CaretIcon');

var _CaretIcon2 = _interopRequireDefault(_CaretIcon);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var cx = _styleHelpers.lucidClassNames.bind('&-SingleSelect');

var any = _propTypes2.default.any,
    bool = _propTypes2.default.bool,
    func = _propTypes2.default.func,
    node = _propTypes2.default.node,
    number = _propTypes2.default.number,
    object = _propTypes2.default.object,
    shape = _propTypes2.default.shape,
    string = _propTypes2.default.string,
    oneOfType = _propTypes2.default.oneOfType;

/**
 *
 * {"categories": ["controls", "selectors"], "madeFrom": ["DropMenu"]}
 *
 * A selector control (like native `<select>`) which is used to select a single option from a dropdown list.
 * Supports option groups with and without labels.
 */

var SingleSelect = (0, _componentTypes.createClass)({
	displayName: 'SingleSelect',

	reducers: reducers,

	components: {
		/**
   * Content this is displayed when nothing is selected.
   */
		Placeholder: (0, _componentTypes.createClass)({
			displayName: 'SingleSelect.Placeholder',
			propName: 'Placeholder'
		}),
		/**
   * A selectable option in the list.
   */
		Option: (0, _componentTypes.createClass)({
			displayName: 'SingleSelect.Option',
			propName: 'Option',
			propTypes: _DropMenu.DropMenuDumb.Option.propTypes
		}),
		/**
   * Groups `Option`s together with a non-selectable heading.
   */
		OptionGroup: (0, _componentTypes.createClass)({
			displayName: 'SingleSelect.OptionGroup',
			propName: 'OptionGroup',
			propTypes: _DropMenu.DropMenuDumb.OptionGroup.propTypes
		})
	},

	propTypes: {
		/**
   * Should be instances of {`SingleSelect.Placeholder`, `SingleSelect.Option`, `SingleSelect.OptionGroup`}. Other direct child elements will not render.
   */
		children: node,
		/**
   * Appended to the component-specific class names set on the root elements. Applies to *both* the control and the flyout menu.
   */
		className: string,
		/**
   * Styles that are passed through to root element.
   */
		style: object,
		/**
   * Applies primary color styling to the control when an item is selected. Defaults to true.
   */
		isSelectionHighlighted: bool,
		/**
   * Allows user to reset the `optionIndex` to `null` if they select the placeholder at the top of the options list.
   * If `false`, it will not render the placeholder in the menu.
   */
		hasReset: bool,
		/**
   * Disables the SingleSelect from being clicked or focused.
   */
		isDisabled: bool,
		/**
   * The currently selected `SingleSelect.Option` index or `null` if nothing is selected.
   */
		selectedIndex: number,
		/**
   * The max height of the fly-out menu.
   */
		maxMenuHeight: oneOfType([number, string]),
		/**
   * Object of DropMenu props which are passed thru to the underlying DropMenu component.
   */
		DropMenu: shape(_DropMenu.DropMenuDumb.propTypes),
		/**
   * Called when an option is selected.
   * Has the signature `(optionIndex, {props, event}) => {}` where `optionIndex` is the new `selectedIndex` or `null` and `props` are the `props` for the selected `Option`.
   */
		onSelect: func,
		/**
   * *Child Element* - The content rendered in the control when there is no option is selected. Also rendered in the option list to remove current selection.
   */
		Placeholder: any,
		/**
   * *Child Element* - These are menu options. The `optionIndex` is in-order of rendering regardless of group nesting, starting with index `0`.
   * Each `Option` may be passed a prop called `isDisabled` to disable selection of that `Option`.
   * Any other props pass to Option will be available from the `onSelect` handler.
   */
		Option: any,
		/**
   * *Child Element* - Used to group `Option`s within the menu. Any non-`Option`s passed in will be rendered as a label for the group.
   */
		OptionGroup: any
	},

	getDefaultProps: function getDefaultProps() {
		return {
			hasReset: true,
			isSelectionHighlighted: true,
			isDisabled: false,
			selectedIndex: null,
			DropMenu: _DropMenu.DropMenuDumb.getDefaultProps()
		};
	},
	getInitialState: function getInitialState() {
		return {
			optionGroups: [],
			flattenedOptionsData: [],
			ungroupedOptionData: [],
			optionGroupDataLookup: {}
		};
	},
	componentWillMount: function componentWillMount() {
		// preprocess the options data before rendering
		this.setState(_DropMenu.DropMenuDumb.preprocessOptionData(this.props, SingleSelect));
	},
	componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
		// only preprocess options data when it changes (via new props) - better performance than doing this each render
		this.setState(_DropMenu.DropMenuDumb.preprocessOptionData(nextProps, SingleSelect));
	},
	render: function render() {
		var _props = this.props,
		    style = _props.style,
		    className = _props.className,
		    hasReset = _props.hasReset,
		    isDisabled = _props.isDisabled,
		    isSelectionHighlighted = _props.isSelectionHighlighted,
		    selectedIndex = _props.selectedIndex,
		    maxMenuHeight = _props.maxMenuHeight,
		    onSelect = _props.onSelect,
		    dropMenuProps = _props.DropMenu;
		var direction = dropMenuProps.direction,
		    isExpanded = dropMenuProps.isExpanded,
		    flyOutStyle = dropMenuProps.flyOutStyle;
		var _state = this.state,
		    optionGroups = _state.optionGroups,
		    optionGroupDataLookup = _state.optionGroupDataLookup,
		    ungroupedOptionData = _state.ungroupedOptionData,
		    flattenedOptionsData = _state.flattenedOptionsData;


		var placeholderProps = (0, _first3.default)((0, _map3.default)((0, _componentTypes.findTypes)(this.props, SingleSelect.Placeholder), 'props'));
		var placeholder = (0, _get3.default)(placeholderProps, 'children', 'Select');
		var isItemSelected = (0, _isNumber3.default)(selectedIndex);

		return _react2.default.createElement(
			_DropMenu.DropMenuDumb,
			_extends({}, dropMenuProps, {
				isDisabled: isDisabled,
				selectedIndices: isItemSelected ? [selectedIndex] : [],
				className: cx('&', className),
				onSelect: onSelect,
				style: style,
				flyOutStyle: (0, _assign3.default)({}, flyOutStyle, !(0, _isNil3.default)(maxMenuHeight) ? { maxHeight: maxMenuHeight } : null)
			}),
			_react2.default.createElement(
				_DropMenu.DropMenuDumb.Control,
				null,
				_react2.default.createElement(
					'div',
					{
						tabIndex: 0,
						className: cx('&-Control', {
							'&-Control-is-highlighted': !isDisabled && isItemSelected && isSelectionHighlighted || isExpanded && isSelectionHighlighted,
							'&-Control-is-selected': !isDisabled && isItemSelected && isSelectionHighlighted || isExpanded && isSelectionHighlighted,
							'&-Control-is-expanded': isExpanded,
							'&-Control-is-disabled': isDisabled
						})
					},
					_react2.default.createElement(
						'span',
						_extends({}, !isItemSelected ? placeholderProps : null, {
							className: cx('&-Control-content', !isItemSelected ? (0, _get3.default)(placeholderProps, 'className') : null)
						}),
						isItemSelected ? flattenedOptionsData[selectedIndex].optionProps.children : placeholder
					),
					_react2.default.createElement(_CaretIcon2.default, { direction: isExpanded ? direction : 'down', size: 8 })
				)
			),
			hasReset && isItemSelected ? _react2.default.createElement(
				_DropMenu.DropMenuDumb.NullOption,
				placeholderProps,
				placeholder
			) : null,
			// for each option group passed in, render a DropMenu.OptionGroup, any label will be included in it's children, render each option inside the group
			(0, _map3.default)(optionGroups, function (optionGroupProps, optionGroupIndex) {
				return _react2.default.createElement(
					_DropMenu.DropMenuDumb.OptionGroup,
					_extends({
						key: 'SingleSelectOptionGroup' + optionGroupIndex
					}, optionGroupProps),
					optionGroupProps.children,
					(0, _map3.default)((0, _get3.default)(optionGroupDataLookup, optionGroupIndex), function (_ref) {
						var optionProps = _ref.optionProps,
						    optionIndex = _ref.optionIndex;
						return _react2.default.createElement(_DropMenu.DropMenuDumb.Option, _extends({
							key: 'SingleSelectOption' + optionIndex
						}, optionProps));
					})
				);
			}).concat(
			// then render all the ungrouped options at the end
			(0, _map3.default)(ungroupedOptionData, function (_ref2) {
				var optionProps = _ref2.optionProps,
				    optionIndex = _ref2.optionIndex;
				return _react2.default.createElement(_DropMenu.DropMenuDumb.Option, _extends({
					key: 'SingleSelectOption' + optionIndex
				}, optionProps));
			}))
		);
	}
});

exports.default = (0, _stateManagement.buildHybridComponent)(SingleSelect);
exports.SingleSelectDumb = SingleSelect;
