// lucid/dist/components/Checkbox/Checkbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _noop2 = require('lodash/noop');

var _noop3 = _interopRequireDefault(_noop2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Checkbox');

var bool = _propTypes2.default.bool,
    func = _propTypes2.default.func,
    object = _propTypes2.default.object,
    string = _propTypes2.default.string;

/**
 * {"categories": ["controls", "toggles"]}
 *
 * This is a checkbox. It's a component that is in one of two particular states
 * at any given moment. It features a custom visualization of the native
 * checkbox button control to reflect its current state.
 *
 * It uses a hidden native checkbox control under the hood but leverages other
 * HTML elements to visualize its state.
 */

var Checkbox = (0, _componentTypes.createClass)({
	displayName: 'Checkbox',
	propTypes: {
		/**
   * Appended to the component-specific class names set on the root
   * element.
   */
		className: string,

		/**
   * Indicates whether the component should appear in an "indeterminate"
   * or "partially checked" state. This prop takes precedence over
   * `isSelected`.
   */
		isIndeterminate: bool,

		/**
   * Indicates whether the component should appear and act disabled by
   * having a "greyed out" palette and ignoring user interactions.
   */
		isDisabled: bool,

		/**
   * Indicates that the component is in the "selected" state when true
   * and in the "unselected" state when false. This props is ignored if
   * `isIndeterminate` is `true`.
   */
		isSelected: bool,

		/**
   * Called when the user clicks on the component or when they press the
   * space key while the component is in focus.
   *
   * Signature: `(isSelected, { event, props }) => {}`
   */
		onSelect: func,

		/**
   * Passed through to the root element.
   */
		style: object
	},

	getDefaultProps: function getDefaultProps() {
		return {
			isIndeterminate: false,
			isDisabled: false,
			isSelected: false,
			onSelect: _noop3.default
		};
	},
	componentDidMount: function componentDidMount() {
		this.nativeElement = this.refs.nativeElement;
	},
	handleSpanClick: function handleSpanClick(e) {
		e.preventDefault();
	},
	render: function render() {
		var _props = this.props,
		    className = _props.className,
		    isIndeterminate = _props.isIndeterminate,
		    isSelected = _props.isSelected,
		    isDisabled = _props.isDisabled,
		    style = _props.style,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'isIndeterminate', 'isSelected', 'isDisabled', 'style']);

		return _react2.default.createElement(
			'div',
			{
				className: cx('&', {
					'&-is-disabled': isDisabled,
					'&-is-selected': isIndeterminate || isSelected
				}, className),
				onClick: this.handleClicked,
				style: style
			},
			_react2.default.createElement('input', _extends({
				onChange: _noop3.default
			}, (0, _componentTypes.omitProps)(passThroughs, Checkbox, ['children']), {
				checked: isSelected,
				className: cx('&-native'),
				disabled: isDisabled,
				ref: 'nativeElement',
				type: 'checkbox'
			})),
			_react2.default.createElement('span', {
				onClick: this.handleSpanClick,
				className: cx('&-visualization-glow')
			}),
			_react2.default.createElement('span', {
				onClick: this.handleSpanClick,
				className: cx('&-visualization-container')
			}),
			isIndeterminate ? _react2.default.createElement(
				'span',
				{
					onClick: this.handleSpanClick,
					className: cx('&-visualization-indeterminate')
				},
				_react2.default.createElement('span', { className: cx('&-visualization-indeterminate-line') })
			) : _react2.default.createElement(
				'span',
				{
					onClick: this.handleSpanClick,
					className: cx('&-visualization-checkmark')
				},
				_react2.default.createElement('span', { className: cx('&-visualization-checkmark-stem') }),
				_react2.default.createElement('span', { className: cx('&-visualization-checkmark-kick') })
			)
		);
	},
	handleClicked: function handleClicked(event) {
		var _props2 = this.props,
		    isDisabled = _props2.isDisabled,
		    isSelected = _props2.isSelected,
		    onSelect = _props2.onSelect;


		if (!isDisabled) {
			onSelect(!isSelected, { event: event, props: this.props });
			this.nativeElement.focus();
		}
	}
});

exports.default = Checkbox;
