// lucid/dist/components/ScrollTable/ScrollTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _componentTypes = require('../../util/component-types');

var _styleHelpers = require('../../util/style-helpers');

var _Table = require('../Table/Table');

var _Table2 = _interopRequireDefault(_Table);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-ScrollTable');

var object = _propTypes2.default.object,
    string = _propTypes2.default.string,
    bool = _propTypes2.default.bool,
    node = _propTypes2.default.node,
    number = _propTypes2.default.number,
    oneOfType = _propTypes2.default.oneOfType;

/**
 * {"categories": ["table"], "madeFrom": ["Table"]}
 *
 * Table in a scrollable container.
 */

var ScrollTable = (0, _componentTypes.createClass)({
	displayName: 'ScrollTable',

	components: {
		Thead: _Table2.default.Thead,
		Tbody: _Table2.default.Tbody,
		Tr: _Table2.default.Tr,
		Th: _Table2.default.Th,
		Td: _Table2.default.Td
	},

	propTypes: {
		/**
   * {Thead, Tbody, Tr, Th, Td} are the child components of Scrolltable, same as Table.
   */
		children: node,
		/**
   * Class names that are appended to the defaults.
   */
		className: string,
		/**
   * Styles that are passed through to the root container.
   */
		style: object,
		/**
   * Set the width of the Table inside the scrollable container.
   */
		tableWidth: oneOfType([number, string]),
		/**
   * Set the Table contents to not allow word wrapping.
   */
		hasWordWrap: bool,
		/**
   * render the table with borders on the outer edge
   */
		hasBorder: bool
	},

	getDefaultProps: function getDefaultProps() {
		return {
			hasWordWrap: false,
			hasBorder: false
		};
	},
	render: function render() {
		var _props = this.props,
		    children = _props.children,
		    className = _props.className,
		    style = _props.style,
		    tableWidth = _props.tableWidth,
		    hasWordWrap = _props.hasWordWrap,
		    hasBorder = _props.hasBorder,
		    passThroughs = _objectWithoutProperties(_props, ['children', 'className', 'style', 'tableWidth', 'hasWordWrap', 'hasBorder']);

		return _react2.default.createElement(
			'div',
			{
				className: cx('&', {
					'&-has-border': hasBorder
				}, className),
				style: style
			},
			_react2.default.createElement(
				_Table2.default,
				_extends({}, (0, _componentTypes.omitProps)(passThroughs, ScrollTable, [], false), {
					style: {
						width: tableWidth
					},
					hasWordWrap: hasWordWrap
				}),
				children
			)
		);
	}
});

exports.default = ScrollTable;
