// lucid/dist/components/SearchableMultiSelect/SearchableMultiSelect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.SearchableMultiSelectDumb = undefined;

var _find2 = require('lodash/find');

var _find3 = _interopRequireDefault(_find2);

var _first2 = require('lodash/first');

var _first3 = _interopRequireDefault(_first2);

var _filter2 = require('lodash/filter');

var _filter3 = _interopRequireDefault(_filter2);

var _isNil2 = require('lodash/isNil');

var _isNil3 = _interopRequireDefault(_isNil2);

var _assign2 = require('lodash/assign');

var _assign3 = _interopRequireDefault(_assign2);

var _isEmpty2 = require('lodash/isEmpty');

var _isEmpty3 = _interopRequireDefault(_isEmpty2);

var _every2 = require('lodash/every');

var _every3 = _interopRequireDefault(_every2);

var _isString2 = require('lodash/isString');

var _isString3 = _interopRequireDefault(_isString2);

var _includes2 = require('lodash/includes');

var _includes3 = _interopRequireDefault(_includes2);

var _omit2 = require('lodash/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _escapeRegExp2 = require('lodash/escapeRegExp');

var _escapeRegExp3 = _interopRequireDefault(_escapeRegExp2);

var _findIndex2 = require('lodash/findIndex');

var _findIndex3 = _interopRequireDefault(_findIndex2);

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _noop2 = require('lodash/noop');

var _noop3 = _interopRequireDefault(_noop2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _stateManagement = require('../../util/state-management');

var _textManipulation = require('../../util/text-manipulation');

var _componentTypes = require('../../util/component-types');

var _SearchField = require('../SearchField/SearchField');

var _DropMenu = require('../DropMenu/DropMenu');

var _LoadingIcon = require('../Icon/LoadingIcon/LoadingIcon');

var _LoadingIcon2 = _interopRequireDefault(_LoadingIcon);

var _Checkbox = require('../Checkbox/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _Selection = require('../Selection/Selection');

var _Selection2 = _interopRequireDefault(_Selection);

var _SearchableMultiSelect = require('./SearchableMultiSelect.reducers');

var reducers = _interopRequireWildcard(_SearchableMultiSelect);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var any = _propTypes2.default.any,
    arrayOf = _propTypes2.default.arrayOf,
    bool = _propTypes2.default.bool,
    func = _propTypes2.default.func,
    number = _propTypes2.default.number,
    oneOfType = _propTypes2.default.oneOfType,
    shape = _propTypes2.default.shape,
    string = _propTypes2.default.string,
    oneOf = _propTypes2.default.oneOf,
    node = _propTypes2.default.node;


var cx = _styleHelpers.lucidClassNames.bind('&-SearchableMultiSelect');

/**
 *
 * {"categories": ["controls", "selectors"], "madeFrom": ["Checkbox", "SearchField", "DropMenu", "LoadingIcon", "Selection"]}
 *
 * A control used to select multiple options from a dropdown list using a
 * SearchField.
 */
var SearchableMultiSelect = (0, _componentTypes.createClass)({
	displayName: 'SearchableMultiSelect',

	reducers: reducers,

	components: {
		/**
   * A selectable option in the list.
   */
		Option: (0, _componentTypes.createClass)({
			displayName: 'SearchableMultiSelect.Option',
			propName: 'Option',
			propTypes: _DropMenu.DropMenuDumb.Option.propTypes,
			components: {
				Selection: (0, _componentTypes.createClass)({
					displayName: 'SearchableMultiSelect.Option.Selection',
					propName: 'Selection',
					propTypes: _Selection2.default.propTypes
				})
			}
		}),
		/**
   * Passes props through to the `SearchField` component.
   */
		SearchField: (0, _componentTypes.createClass)({
			displayName: 'SearchableMultiSelect.SearchField',
			propName: 'SearchField',
			propTypes: _SearchField.SearchFieldDumb.propTypes
		}),

		/**
   * Groups `Option`s together with a non-selectable heading.
   */
		OptionGroup: (0, _componentTypes.createClass)({
			displayName: 'SearchableMultiSelect.OptionGroup',
			propName: 'OptionGroup',
			propTypes: _DropMenu.DropMenuDumb.OptionGroup.propTypes
		}),

		/**
   * Label for the selected section header.
   */
		SelectionLabel: (0, _componentTypes.createClass)({
			displayName: 'SearchableMultiSelect.SelectionLabel',
			propName: 'SelectionLabel'
		})
	},

	propTypes: {
		/**
   * Should be instances of {`SearchableMultiSelect.Option`}. Other direct child elements will not render.
   */
		children: node,
		/**
   * Appended to the component-specific class names set on the root element.
   */
		className: string,
		/**
   * Disables the control from being clicked or focused.
   */
		isDisabled: bool,
		/**
   * Displays a LoadingIcon to allow for asynchronous loading of options.
   */
		isLoading: bool,
		/**
   * The max height of the fly-out menu.
   */
		maxMenuHeight: oneOfType([number, string]),
		/**
   * Called when the user enters a value to search for; the set of visible
   * Options will be filtered using the value.
   *
   * Signature: `(searchText, firstVisibleIndex, {props, event}) => {}`
   *
   * `searchText` is the value from the `SearchField` and `firstVisibleIndex`
   * is the index of the first option that will be visible after filtering.
   */
		onSearch: func,
		/**
   * Called when an option is selected.
   *
   * Signature: `(optionIndex, {props, event}) => {}`
   *
   * `optionIndex` is the new `selectedIndex` or `null`.
   */
		onSelect: func,
		/**
   * Called when the user clicks to remove all selections.
   *
   * Signature: `({props, event}) => {}`
   */
		onRemoveAll: func,
		/**
   * The function that will be run against each Option's props to determine
   * whether it should be visible or not. The default behavior of the
   * function is to match, ignoring case, against any text node descendant of
   * the `Option`.
   *
   * Signature: `(searchText, optionProps) => {}`
   *
   * If `true` is returned, the option will be visible. If `false`, the
   * option will not be visible.
   */
		optionFilter: func,
		/**
   * The current search text to filter the list of options by.
   */
		searchText: string,
		/**
   * An array of currently selected `SearchableMultiSelect.Option` indices or
   * `null` if nothing is selected.
   */
		selectedIndices: arrayOf(number),
		/**
   * Object of DropMenu props which are passed through to the underlying
   * DropMenu component.
   */
		DropMenu: shape(_DropMenu.DropMenuDumb.propTypes),
		/**
   * *Child Element* - These are menu options. Each `Option` may be passed a
   * prop called `isDisabled` to disable selection of that `Option`. Any
   * other props pass to Option will be available from the `onSelect`
   * handler.
   *
   * It also support the `Selection` prop that can be used to forward along
   * props to the underlying `Selection` component.
   */
		Option: any,
		/**
   * Adjusts the display of this component. This should typically be driven
   * by screen size. Currently `small` and `large` are explicitly handled
   * by this component.
   */
		responsiveMode: oneOf(['small', 'medium', 'large']),
		/**
   * Controls the visibility of the "remove all" button that's shown with the
   * selected items.
   */
		hasRemoveAll: bool,
		/**
   * Controls the visibility of the `Selection` component that appears below
   * the search field.
   */
		hasSelections: bool
	},

	getInitialState: function getInitialState() {
		return {
			optionGroups: [],
			flattenedOptionsData: [],
			ungroupedOptionData: [],
			optionGroupDataLookup: {}
		};
	},
	getDefaultProps: function getDefaultProps() {
		return {
			isDisabled: false,
			isLoading: false,
			onRemoveAll: _noop3.default,
			optionFilter: _textManipulation.propsSearch,
			searchText: '',
			selectedIndices: [],
			DropMenu: _DropMenu.DropMenuDumb.getDefaultProps(),
			responsiveMode: 'large',
			hasRemoveAll: true,
			hasSelections: true
		};
	},
	handleDropMenuSelect: function handleDropMenuSelect(optionIndex, _ref) {
		var event = _ref.event,
		    props = _ref.props;
		var onSelect = this.props.onSelect;


		return onSelect(optionIndex, { event: event, props: props });
	},
	handleCheckboxSelect: function handleCheckboxSelect(_isSelected, _ref2) {
		var event = _ref2.event,
		    optionIndex = _ref2.props.callbackId;

		// This is needed otherwise clicking the checkbox will double fire this
		// event _and_ the `handleDropMenuSelect` handler
		event.stopPropagation();

		// We don't want to send the consumer the checkbox's props so we have to
		// lookup the option they clicked and send its props along
		var selectedOptionProps = (0, _get3.default)((0, _componentTypes.findTypes)(this.props, SearchableMultiSelect.Option), '[' + optionIndex + '].props');

		return this.props.onSelect(optionIndex, {
			event: event,
			props: selectedOptionProps
		});
	},
	handleSelectionRemove: function handleSelectionRemove(_ref3) {
		var event = _ref3.event,
		    props = _ref3.props,
		    optionIndex = _ref3.props.callbackId;

		// We don't want to send the consumer the selection's props so we have to
		// lookup the option they clicked and send its props along
		var selectedOptionProps = (0, _get3.default)((0, _componentTypes.findTypes)(this.props, SearchableMultiSelect.Option), '[' + optionIndex + '].props');

		return this.props.onSelect(optionIndex, {
			event: event,
			props: selectedOptionProps
		});
	},
	handleRemoveAll: function handleRemoveAll(_ref4) {
		var event = _ref4.event;

		this.props.onRemoveAll({ event: event, props: this.props });
	},
	handleSearch: function handleSearch(searchText, _ref5) {
		var event = _ref5.event;
		var props = this.props,
		    _props = this.props,
		    onSearch = _props.onSearch,
		    optionFilter = _props.optionFilter,
		    onExpand = _props.DropMenu.onExpand;


		var options = (0, _map3.default)((0, _componentTypes.findTypes)(props, SearchableMultiSelect.Option), 'props');
		var firstVisibleIndex = (0, _findIndex3.default)(options, function (option) {
			return optionFilter(searchText, option);
		});
		var firstVisibleProps = options[firstVisibleIndex];

		// Just an extra call to make sure the search results show up when a user
		// is typing
		onExpand();

		return onSearch(searchText, firstVisibleIndex, {
			event: event,
			props: firstVisibleProps
		});
	},
	componentWillMount: function componentWillMount() {
		// preprocess the options data before rendering
		this.setState(_DropMenu.DropMenuDumb.preprocessOptionData(this.props, SearchableMultiSelect));
	},
	componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
		// only preprocess options data when it changes (via new props) - better performance than doing this each render
		this.setState(_DropMenu.DropMenuDumb.preprocessOptionData(nextProps, SearchableMultiSelect));
	},
	renderUnderlinedChildren: function renderUnderlinedChildren(childText, searchText) {
		var _partitionText = (0, _textManipulation.partitionText)(childText, new RegExp((0, _escapeRegExp3.default)(searchText), 'i'), searchText.length),
		    _partitionText2 = _slicedToArray(_partitionText, 3),
		    pre = _partitionText2[0],
		    match = _partitionText2[1],
		    post = _partitionText2[2];

		return [pre && _react2.default.createElement(
			'span',
			{ key: 'pre', className: cx('&-Option-underline-pre') },
			pre
		), match && _react2.default.createElement(
			'span',
			{ key: 'match', className: cx('&-Option-underline-match') },
			match
		), post && _react2.default.createElement(
			'span',
			{ key: 'post', className: cx('&-Option-underline-post') },
			post
		)];
	},
	renderOption: function renderOption(_ref6) {
		var optionProps = _ref6.optionProps,
		    optionIndex = _ref6.optionIndex;
		var _props2 = this.props,
		    optionFilter = _props2.optionFilter,
		    searchText = _props2.searchText,
		    selectedIndices = _props2.selectedIndices,
		    isLoading = _props2.isLoading;


		return _react2.default.createElement(
			_DropMenu.DropMenuDumb.Option,
			_extends({
				key: 'SearchableMultiSelectOption' + optionIndex
			}, (0, _omit3.default)(optionProps, 'children'), {
				isHidden: !optionFilter(searchText, optionProps),
				isDisabled: isLoading
			}),
			_react2.default.createElement(
				'div',
				{ className: cx('&-checkbox') },
				_react2.default.createElement(_Checkbox2.default, {
					onSelect: this.handleCheckboxSelect,
					callbackId: optionIndex,
					isSelected: (0, _includes3.default)(selectedIndices, optionIndex)
				}),
				_react2.default.createElement(
					'label',
					{ className: cx('&-checkbox-label') },
					(0, _isString3.default)(optionProps.children) ? this.renderUnderlinedChildren(optionProps.children, searchText) : optionProps.children
				)
			)
		);
	},
	renderOptions: function renderOptions() {
		var _this = this;

		var _props3 = this.props,
		    optionFilter = _props3.optionFilter,
		    searchText = _props3.searchText,
		    isLoading = _props3.isLoading;
		var _state = this.state,
		    optionGroups = _state.optionGroups,
		    optionGroupDataLookup = _state.optionGroupDataLookup,
		    ungroupedOptionData = _state.ungroupedOptionData,
		    flattenedOptionsData = _state.flattenedOptionsData;


		var isAllOptionsHidden = (0, _every3.default)(flattenedOptionsData, function (_ref7) {
			var optionProps = _ref7.optionProps;
			return !optionFilter(searchText, optionProps);
		});

		// for each option group passed in, render a DropMenu.OptionGroup, any label will be included in it's children, render each option inside the group
		var dropMenuOptions = (0, _map3.default)(optionGroups, function (optionGroupProps, optionGroupIndex) {
			return _react2.default.createElement(
				_DropMenu.DropMenuDumb.OptionGroup,
				_extends({
					key: 'SearchableMultiSelectOptionGroup' + optionGroupIndex
				}, (0, _omit3.default)(optionGroupProps, 'children')),
				optionGroupProps.children,
				(0, _map3.default)(optionGroupDataLookup[optionGroupIndex], _this.renderOption)
			);
		}).concat(
		// then render all the ungrouped options at the end
		(0, _map3.default)(ungroupedOptionData, this.renderOption));

		if (!isAllOptionsHidden || (0, _isEmpty3.default)(searchText)) {
			return dropMenuOptions;
		}

		if (!isLoading) {
			return _react2.default.createElement(
				_DropMenu.DropMenuDumb.Option,
				{ isDisabled: true },
				_react2.default.createElement(
					'span',
					{ className: cx('&-noresults') },
					'No results match "',
					searchText,
					'"'
				)
			);
		}

		return null;
	},
	render: function render() {
		var _this2 = this;

		var props = this.props,
		    _props4 = this.props,
		    className = _props4.className,
		    isLoading = _props4.isLoading,
		    isDisabled = _props4.isDisabled,
		    maxMenuHeight = _props4.maxMenuHeight,
		    selectedIndices = _props4.selectedIndices,
		    dropMenuProps = _props4.DropMenu,
		    optionContainerStyle = _props4.DropMenu.optionContainerStyle,
		    responsiveMode = _props4.responsiveMode,
		    searchText = _props4.searchText,
		    hasRemoveAll = _props4.hasRemoveAll,
		    hasSelections = _props4.hasSelections,
		    passThroughs = _objectWithoutProperties(_props4, ['className', 'isLoading', 'isDisabled', 'maxMenuHeight', 'selectedIndices', 'DropMenu', 'DropMenu', 'responsiveMode', 'searchText', 'hasRemoveAll', 'hasSelections']);

		var _state2 = this.state,
		    optionGroupDataLookup = _state2.optionGroupDataLookup,
		    optionGroups = _state2.optionGroups,
		    ungroupedOptionData = _state2.ungroupedOptionData;


		var searchFieldProps = (0, _get3.default)((0, _componentTypes.getFirst)(props, SearchableMultiSelect.SearchField), 'props', {});
		var selectionLabel = (0, _componentTypes.getFirst)(props, SearchableMultiSelect.SelectionLabel, _react2.default.createElement(
			SearchableMultiSelect.SelectionLabel,
			null,
			'Selected'
		));
		var isSmall = responsiveMode === 'small';

		return _react2.default.createElement(
			'div',
			_extends({}, (0, _componentTypes.omitProps)(passThroughs, SearchableMultiSelect), {
				className: cx('&', className)
			}),
			_react2.default.createElement(
				_DropMenu.DropMenuDumb,
				_extends({}, dropMenuProps, {
					selectedIndices: null,
					className: cx('&-DropMenu', {
						'&-DropMenu-is-small': isSmall
					}, dropMenuProps.className),
					optionContainerStyle: (0, _assign3.default)({}, optionContainerStyle, !(0, _isNil3.default)(maxMenuHeight) ? { maxHeight: maxMenuHeight } : null),
					isDisabled: isDisabled,
					isLoading: isLoading,
					onSelect: this.handleDropMenuSelect,
					ContextMenu: {
						alignmentOffset: -13,
						directonOffset: -1,
						minWidthOffset: -28
					}
				}),
				_react2.default.createElement(
					_DropMenu.DropMenuDumb.Control,
					null,
					_react2.default.createElement(_SearchField.SearchFieldDumb, _extends({}, searchFieldProps, {
						isDisabled: isDisabled,
						className: cx('&-search', {
							'&-search-is-small': isSmall
						}, searchFieldProps.className),
						value: searchText,
						onChange: this.handleSearch
					}))
				),
				isLoading ? _react2.default.createElement(
					_DropMenu.DropMenuDumb.Option,
					{
						key: 'SearchableMultiSelectLoading',
						className: cx('&-loading'),
						isDisabled: true
					},
					_react2.default.createElement(_LoadingIcon2.default, null)
				) : null,
				this.renderOptions()
			),
			hasSelections && !(0, _isEmpty3.default)(selectedIndices) ? _react2.default.createElement(
				_Selection2.default,
				{
					className: cx('&-Selection-section'),
					isBold: true,
					hasBackground: true,
					kind: 'container',
					onRemove: this.handleRemoveAll,
					responsiveMode: responsiveMode,
					isRemovable: hasRemoveAll
				},
				_react2.default.createElement(
					_Selection2.default.Label,
					null,
					selectionLabel.props.children
				),
				(0, _map3.default)(optionGroupDataLookup, function (groupedOptionsData, optionGroupIndex) {
					var selectedGroupedOptions = (0, _filter3.default)(groupedOptionsData, function (_ref8) {
						var optionIndex = _ref8.optionIndex;
						return (0, _includes3.default)(selectedIndices, optionIndex);
					});
					if (!(0, _isEmpty3.default)(selectedGroupedOptions)) {
						return _react2.default.createElement(
							_Selection2.default,
							{
								className: cx('&-Selection-group'),
								key: 'optionGroup-' + optionGroupIndex,
								responsiveMode: responsiveMode,
								isRemovable: false,
								isBold: true,
								kind: 'container'
							},
							_react2.default.createElement(
								_Selection2.default.Label,
								null,
								(0, _first3.default)((0, _componentTypes.rejectTypes)(optionGroups[optionGroupIndex].children, SearchableMultiSelect.Option))
							),
							(0, _map3.default)(selectedGroupedOptions, function (_ref9) {
								var optionIndex = _ref9.optionIndex,
								    optionProps = _ref9.optionProps;

								var selectionProps = (0, _get3.default)((0, _componentTypes.getFirst)(optionProps, SearchableMultiSelect.Option.Selection), 'props');
								return _react2.default.createElement(
									_Selection2.default,
									_extends({
										key: optionIndex
									}, selectionProps, {
										callbackId: optionIndex,
										responsiveMode: responsiveMode,
										onRemove: _this2.handleSelectionRemove
									}),
									_react2.default.createElement(
										_Selection2.default.Label,
										null,
										optionProps.children
									)
								);
							})
						);
					}
					return null;
				}),
				(0, _map3.default)(selectedIndices, function (selectedIndex) {
					var selectedUngroupedOptionData = (0, _find3.default)(ungroupedOptionData, { optionIndex: selectedIndex });

					if (selectedUngroupedOptionData) {
						var optionProps = selectedUngroupedOptionData.optionProps;

						var selectionProps = (0, _get3.default)((0, _componentTypes.getFirst)(optionProps, SearchableMultiSelect.Option.Selection), 'props');
						return _react2.default.createElement(
							_Selection2.default,
							_extends({
								key: selectedIndex
							}, selectionProps, {
								callbackId: selectedIndex,
								responsiveMode: responsiveMode,
								onRemove: _this2.handleSelectionRemove
							}),
							_react2.default.createElement(
								_Selection2.default.Label,
								null,
								optionProps.children
							)
						);
					}
					return null;
				})
			) : null
		);
	}
});

exports.default = (0, _stateManagement.buildHybridComponent)(SearchableMultiSelect);
exports.SearchableMultiSelectDumb = SearchableMultiSelect;
