// lucid/dist/components/SearchableMultiSelect/SearchableMultiSelect.reducers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.SearchField = exports.DropMenu = undefined;

var _xor2 = require('lodash/xor');

var _xor3 = _interopRequireDefault(_xor2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.onSelect = onSelect;
exports.onSearch = onSearch;
exports.onRemoveAll = onRemoveAll;

var _DropMenu = require('../DropMenu/DropMenu.reducers');

var DropMenu = _interopRequireWildcard(_DropMenu);

var _SearchField = require('../SearchField/SearchField.reducers');

var _SearchField2 = _interopRequireDefault(_SearchField);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function onSelect(state, selectedIndex) {
	return _extends({}, state, {
		selectedIndices: (0, _xor3.default)(state.selectedIndices, [selectedIndex])
	});
}

function onSearch(state, searchText, firstVisibleIndex) {
	return _extends({}, state, {
		searchText: searchText,
		DropMenu: _extends({}, DropMenu.onFocusOption(state.DropMenu, firstVisibleIndex))
	});
}

function onRemoveAll(state) {
	return _extends({}, state, {
		selectedIndices: []
	});
}

exports.DropMenu = DropMenu;
exports.SearchField = _SearchField2.default;
