// lucid/dist/components/Overlay/Overlay.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _uniqueId2 = require('lodash/uniqueId');

var _uniqueId3 = _interopRequireDefault(_uniqueId2);

var _noop2 = require('lodash/noop');

var _noop3 = _interopRequireDefault(_noop2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Portal = require('../Portal/Portal');

var _Portal2 = _interopRequireDefault(_Portal);

var _CSSTransitionGroup = require('react-transition-group/CSSTransitionGroup');

var _CSSTransitionGroup2 = _interopRequireDefault(_CSSTransitionGroup);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Overlay');

var string = _propTypes2.default.string,
    bool = _propTypes2.default.bool,
    func = _propTypes2.default.func,
    node = _propTypes2.default.node;

/**
 * {"categories": ["utility"], "madeFrom": ["Portal"]}
 *
 * Overlay is used to block user interaction with the rest of the app until they
 * have completed something.
 */

var Overlay = (0, _componentTypes.createClass)({
	displayName: 'Overlay',

	propTypes: {
		/**
   * Appended to the component-specific class names set on the root element.
   */
		className: string,

		/**
   * Generally you should only have a single child element so the centering
   * works correctly.
   */
		children: node,

		/**
   * Controls visibility.
   */
		isShown: bool,

		/**
   * Determines if it shows with a gray background. If `false`, the
   * background will be rendered but will be invisible, except for the
   * contents, and it won't capture any of the user click events.
   */
		isModal: bool,

		/**
   * Set your own id for the `Portal` is that is opened up to contain the
   * contents. In practice you should never need to set this manually.
   */
		portalId: string,

		/**
   * Fired when the user hits escape.
   *
   * Signature: `({ event, props }) => {}`
   */
		onEscape: func,

		/**
   * Fired when the user clicks on the background, this may or may not be
   * visible depending on `isModal`.
   *
   * Signature: `({ event, props }) => {}`
   */
		onBackgroundClick: func
	},

	getDefaultProps: function getDefaultProps() {
		return {
			isShown: false,
			isModal: true,
			onEscape: _noop3.default,
			onBackgroundClick: _noop3.default
		};
	},
	getInitialState: function getInitialState() {
		return {
			// This must be in state because getDefaultProps only runs once per
			// component import which causes collisions
			portalId: this.props.portalId || (0, _uniqueId3.default)('Overlay-Portal-')
		};
	},
	componentDidMount: function componentDidMount() {
		if (window && window.document) {
			window.document.addEventListener('keydown', this.handleDocumentKeyDown);
		}
	},
	componentWillUnmount: function componentWillUnmount() {
		if (window && window.document) {
			window.document.removeEventListener('keydown', this.handleDocumentKeyDown);
		}
	},
	handleDocumentKeyDown: function handleDocumentKeyDown(event) {
		// If the user hits the "escape" key, then fire an `onEscape`
		// TODO: use key helpers
		if (event.keyCode === 27) {
			this.props.onEscape({ event: event, props: this.props });
		}
	},
	handleDivRef: function handleDivRef(divDOMNode) {
		// Store the dom node so we can check if it's clicked on later
		this._divDOMNode = _reactDom2.default.findDOMNode(divDOMNode);
	},
	handleBackgroundClick: function handleBackgroundClick(event) {
		// Use the reference we previously stored from the `ref` to check what
		// element was clicked on.
		if (this._divDOMNode && event.target === this._divDOMNode) {
			this.props.onBackgroundClick({ event: event, props: this.props });
		}
	},
	render: function render() {
		var _props = this.props,
		    className = _props.className,
		    isShown = _props.isShown,
		    isModal = _props.isModal,
		    children = _props.children,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'isShown', 'isModal', 'children']);

		var portalId = this.state.portalId;


		return _react2.default.createElement(
			_Portal2.default,
			{ portalId: portalId },
			_react2.default.createElement(
				_CSSTransitionGroup2.default,
				{
					transitionName: cx('&'),
					transitionEnterTimeout: 300,
					transitionLeaveTimeout: 300
				},
				isShown ? _react2.default.createElement(
					'div',
					_extends({}, (0, _componentTypes.omitProps)(passThroughs, Overlay), {
						className: cx(className, '&', {
							'&-is-not-modal': !isModal
						}),
						onClick: this.handleBackgroundClick,
						ref: this.handleDivRef
					}),
					children
				) : null
			)
		);
	}
});

exports.default = Overlay;
