// lucid/dist/constants/charts.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
// NOTE: these needs to stay in sync with variables.less

// Each of these map to a less variable that applies the actual color
var COLOR_0 = exports.COLOR_0 = 'color-chart-0';
var COLOR_0_LIGHTEST = exports.COLOR_0_LIGHTEST = 'color-chart-0-lightest';
var COLOR_0_LIGHT = exports.COLOR_0_LIGHT = 'color-chart-0-light';
var COLOR_0_DARK = exports.COLOR_0_DARK = 'color-chart-0-dark';
var COLOR_0_DARKEST = exports.COLOR_0_DARKEST = 'color-chart-0-darkest';
var COLOR_1 = exports.COLOR_1 = 'color-chart-1';
var COLOR_1_LIGHTEST = exports.COLOR_1_LIGHTEST = 'color-chart-1-lightest';
var COLOR_1_LIGHT = exports.COLOR_1_LIGHT = 'color-chart-1-light';
var COLOR_1_DARK = exports.COLOR_1_DARK = 'color-chart-1-dark';
var COLOR_1_DARKEST = exports.COLOR_1_DARKEST = 'color-chart-1-darkest';
var COLOR_2 = exports.COLOR_2 = 'color-chart-2';
var COLOR_2_LIGHTEST = exports.COLOR_2_LIGHTEST = 'color-chart-2-lightest';
var COLOR_2_LIGHT = exports.COLOR_2_LIGHT = 'color-chart-2-light';
var COLOR_2_DARK = exports.COLOR_2_DARK = 'color-chart-2-dark';
var COLOR_2_DARKEST = exports.COLOR_2_DARKEST = 'color-chart-2-darkest';
var COLOR_3 = exports.COLOR_3 = 'color-chart-3';
var COLOR_3_LIGHTEST = exports.COLOR_3_LIGHTEST = 'color-chart-3-lightest';
var COLOR_3_LIGHT = exports.COLOR_3_LIGHT = 'color-chart-3-light';
var COLOR_3_DARK = exports.COLOR_3_DARK = 'color-chart-3-dark';
var COLOR_3_DARKEST = exports.COLOR_3_DARKEST = 'color-chart-3-darkest';
var COLOR_4 = exports.COLOR_4 = 'color-chart-4';
var COLOR_4_LIGHTEST = exports.COLOR_4_LIGHTEST = 'color-chart-4-lightest';
var COLOR_4_LIGHT = exports.COLOR_4_LIGHT = 'color-chart-4-light';
var COLOR_4_DARK = exports.COLOR_4_DARK = 'color-chart-4-dark';
var COLOR_4_DARKEST = exports.COLOR_4_DARKEST = 'color-chart-4-darkest';
var COLOR_5 = exports.COLOR_5 = 'color-chart-5';
var COLOR_5_LIGHTEST = exports.COLOR_5_LIGHTEST = 'color-chart-5-lightest';
var COLOR_5_LIGHT = exports.COLOR_5_LIGHT = 'color-chart-5-light';
var COLOR_5_DARK = exports.COLOR_5_DARK = 'color-chart-5-dark';
var COLOR_5_DARKEST = exports.COLOR_5_DARKEST = 'color-chart-5-darkest';
var COLOR_GOOD = exports.COLOR_GOOD = 'color-chart-good';
var COLOR_GOOD_LIGHTEST = exports.COLOR_GOOD_LIGHTEST = 'color-chart-good-lightest';
var COLOR_GOOD_LIGHT = exports.COLOR_GOOD_LIGHT = 'color-chart-good-light';
var COLOR_GOOD_DARK = exports.COLOR_GOOD_DARK = 'color-chart-good-dark';
var COLOR_GOOD_DARKEST = exports.COLOR_GOOD_DARKEST = 'color-chart-good-darkest';
var COLOR_BAD = exports.COLOR_BAD = 'color-chart-bad';
var COLOR_BAD_LIGHTEST = exports.COLOR_BAD_LIGHTEST = 'color-chart-bad-lightest';
var COLOR_BAD_LIGHT = exports.COLOR_BAD_LIGHT = 'color-chart-bad-light';
var COLOR_BAD_DARK = exports.COLOR_BAD_DARK = 'color-chart-bad-dark';
var COLOR_BAD_DARKEST = exports.COLOR_BAD_DARKEST = 'color-chart-bad-darkest';
var COLOR_NEUTRAL = exports.COLOR_NEUTRAL = 'color-chart-neutral';

var PALETTE_6 = exports.PALETTE_6 = [COLOR_0, COLOR_1, COLOR_2, COLOR_3, COLOR_4, COLOR_5];

var PALETTE_30 = exports.PALETTE_30 = [COLOR_0_LIGHTEST, COLOR_0_LIGHT, COLOR_0, COLOR_0_DARK, COLOR_0_DARKEST, COLOR_1_LIGHTEST, COLOR_1_LIGHT, COLOR_1, COLOR_1_DARK, COLOR_1_DARKEST, COLOR_2_LIGHTEST, COLOR_2_LIGHT, COLOR_2, COLOR_2_DARK, COLOR_2_DARKEST, COLOR_3_LIGHTEST, COLOR_3_LIGHT, COLOR_3, COLOR_3_DARK, COLOR_3_DARKEST, COLOR_4_LIGHTEST, COLOR_4_LIGHT, COLOR_4, COLOR_4_DARK, COLOR_4_DARKEST, COLOR_5_LIGHTEST, COLOR_5_LIGHT, COLOR_5, COLOR_5_DARK, COLOR_5_DARKEST];

var PALETTE_MONOCHROME_GOOD_3 = exports.PALETTE_MONOCHROME_GOOD_3 = [COLOR_GOOD_LIGHT, COLOR_GOOD, COLOR_GOOD_DARK];

var PALETTE_MONOCHROME_GOOD_5 = exports.PALETTE_MONOCHROME_GOOD_5 = [COLOR_GOOD_LIGHTEST, COLOR_GOOD_LIGHT, COLOR_GOOD, COLOR_GOOD_DARK, COLOR_GOOD_DARKEST];

var PALETTE_MONOCHROME_BAD_3 = exports.PALETTE_MONOCHROME_BAD_3 = [COLOR_BAD_LIGHT, COLOR_BAD, COLOR_BAD_DARK];

var PALETTE_MONOCHROME_BAD_5 = exports.PALETTE_MONOCHROME_BAD_5 = [COLOR_BAD_LIGHTEST, COLOR_BAD_LIGHT, COLOR_BAD, COLOR_BAD_DARK, COLOR_BAD_DARKEST];

var PALETTE_MONOCHROME_0_5 = exports.PALETTE_MONOCHROME_0_5 = [COLOR_0_LIGHTEST, COLOR_0_LIGHT, COLOR_0, COLOR_0_DARK, COLOR_0_DARKEST];

var PALETTE_MONOCHROME_1_5 = exports.PALETTE_MONOCHROME_1_5 = [COLOR_1_LIGHTEST, COLOR_1_LIGHT, COLOR_1, COLOR_1_DARK, COLOR_1_DARKEST];

var PALETTE_MONOCHROME_2_5 = exports.PALETTE_MONOCHROME_2_5 = [COLOR_2_LIGHTEST, COLOR_2_LIGHT, COLOR_2, COLOR_2_DARK, COLOR_2_DARKEST];

var PALETTE_MONOCHROME_3_5 = exports.PALETTE_MONOCHROME_3_5 = [COLOR_3_LIGHTEST, COLOR_3_LIGHT, COLOR_3, COLOR_3_DARK, COLOR_3_DARKEST];

var PALETTE_MONOCHROME_4_5 = exports.PALETTE_MONOCHROME_4_5 = [COLOR_4_LIGHTEST, COLOR_4_LIGHT, COLOR_4, COLOR_4_DARK, COLOR_4_DARKEST];

var PALETTE_MONOCHROME_5_5 = exports.PALETTE_MONOCHROME_5_5 = [COLOR_5_LIGHTEST, COLOR_5_LIGHT, COLOR_5, COLOR_5_DARK, COLOR_5_DARKEST];
