// lucid/dist/components/Badge/Badge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Badge');

var node = _propTypes2.default.node,
    string = _propTypes2.default.string;

/**
 *
 * {"categories": ["visual design", "icons"]}
 *
 * `Badge` is a quick utility component to create a badge around any
 * element(s). Do not wrap existing `Icon`s in a badge, rather add the
 * `isBadge` prop to any Icon component to turn it into a badge.
 */

var Badge = (0, _componentTypes.createClass)({
	displayName: 'Badge',
	propTypes: {
		/**
   * class names that are appended to the defaults
   */
		className: string,
		/**
   * any valid React children
   */
		children: node
	},

	render: function render() {
		var _props = this.props,
		    className = _props.className,
		    children = _props.children,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'children']);

		return _react2.default.createElement(
			'span',
			_extends({ className: cx('&', className) }, (0, _componentTypes.omitProps)(passThroughs, Badge)),
			children
		);
	}
});

exports.default = Badge;
