// lucid/dist/components/Autocomplete/Autocomplete.reducers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.DropMenu = undefined;

var _isEmpty2 = require('lodash/isEmpty');

var _isEmpty3 = _interopRequireDefault(_isEmpty2);

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.onChange = onChange;
exports.onSelect = onSelect;
exports.onExpand = onExpand;

var _DropMenu = require('../DropMenu/DropMenu.reducers');

var DropMenu = _interopRequireWildcard(_DropMenu);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function onChange(state, value) {
	return _extends({}, state, {
		value: value,
		DropMenu: _extends({}, state.DropMenu, {
			focusedIndex: null
		})
	});
}

function onSelect(state, selectedIndex) {
	var value = (0, _get3.default)(state.suggestions, selectedIndex);

	return onChange(_extends({}, state, {
		value: value,
		DropMenu: _extends({}, DropMenu.onSelect(state.DropMenu, selectedIndex), {
			selectedIndices: []
		})
	}), value);
}

function onExpand(state) {
	return _extends({}, state, {
		DropMenu: _extends({}, state.DropMenu, {
			focusedIndex: null,
			isExpanded: !(0, _isEmpty3.default)(state.value) && !(0, _isEmpty3.default)(state.suggestions)
		})
	});
}

exports.DropMenu = DropMenu;
