// lucid/dist/components/SwitchLabeled/SwitchLabeled.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _noop2 = require('lodash/noop');

var _noop3 = _interopRequireDefault(_noop2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _CSSTransitionGroup = require('react-transition-group/CSSTransitionGroup');

var _CSSTransitionGroup2 = _interopRequireDefault(_CSSTransitionGroup);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

var _Switch = require('../Switch/Switch');

var _Switch2 = _interopRequireDefault(_Switch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-SwitchLabeled');
var any = _propTypes2.default.any,
    node = _propTypes2.default.node,
    object = _propTypes2.default.object,
    string = _propTypes2.default.string;

/**
 * {"categories": ["controls", "toggles"], "extend": "Switch", "madeFrom": ["Switch"]}
 *
 * This is a composite of the `Switch` component and the native `label`
 * element.
 */

var SwitchLabeled = (0, _componentTypes.createClass)({
	displayName: 'SwitchLabeled',

	components: {
		/**
   * Label to be shown alongside the switch.
   */
		Label: (0, _componentTypes.createClass)({
			displayName: 'SwitchLabeled.Label',
			propName: 'Label',
			propTypes: {
				/**
     * Used to identify the purpose of this switch to the user -- can be
     * any renderable content.
     */
				children: node
			}
		})
	},

	propTypes: _extends({}, _Switch2.default.propTypes, {

		/**
   * Appended to the component-specific class names set on the root
   * element.
   */
		className: string,

		/**
   * Passed through to the root element.
   */
		style: object,

		/**
   * Child element whose children are used to identify the purpose of this
   * switch to the user.
   */
		Label: any
	}),

	getDefaultProps: function getDefaultProps() {
		return {
			isDisabled: false,
			isSelected: false,
			onSelect: _noop3.default
		};
	},
	componentWillMount: function componentWillMount() {
		this._labelKey = 0;
	},
	componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
		var currentLabel = (0, _get3.default)((0, _componentTypes.getFirst)(this.props, SwitchLabeled.Label), 'props.children', null);
		var nextLabel = (0, _get3.default)((0, _componentTypes.getFirst)(nextProps, SwitchLabeled.Label), 'props.children', null);

		if (currentLabel !== nextLabel) {
			this._labelKey++;
		}
	},
	render: function render() {
		var _props = this.props,
		    className = _props.className,
		    isDisabled = _props.isDisabled,
		    isSelected = _props.isSelected,
		    onSelect = _props.onSelect,
		    style = _props.style,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'isDisabled', 'isSelected', 'onSelect', 'style']);

		var labelChildProps = (0, _get3.default)((0, _componentTypes.getFirst)(this.props, SwitchLabeled.Label), 'props');

		return _react2.default.createElement(
			'label',
			{
				className: cx('&', {
					'&-is-disabled': isDisabled,
					'&-is-selected': isSelected
				}, className),
				style: style
			},
			_react2.default.createElement(_Switch2.default, _extends({
				className: className,
				isDisabled: isDisabled,
				isSelected: isSelected,
				onSelect: onSelect
			}, (0, _componentTypes.omitProps)(passThroughs, SwitchLabeled, [], false))),
			_react2.default.createElement(
				_CSSTransitionGroup2.default,
				{
					transitionName: cx('&-text'),
					transitionEnterTimeout: 100,
					transitionLeaveTimeout: 100,
					style: { position: 'relative' },
					className: cx('&-text')
				},
				labelChildProps ? _react2.default.createElement(
					'span',
					{ key: this._labelKey },
					labelChildProps.children || labelChildProps
				) : null
			)
		);
	}
});

exports.default = SwitchLabeled;
