// lucid/dist/components/EmptyStateWrapper/EmptyStateWrapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _componentTypes = require('../../util/component-types');

var _styleHelpers = require('../../util/style-helpers');

var _LoadingIndicator = require('../LoadingIndicator/LoadingIndicator');

var _LoadingIndicator2 = _interopRequireDefault(_LoadingIndicator);

var _OverlayWrapper = require('../OverlayWrapper/OverlayWrapper');

var _OverlayWrapper2 = _interopRequireDefault(_OverlayWrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-EmptyStateWrapper');

var any = _propTypes2.default.any,
    bool = _propTypes2.default.bool,
    node = _propTypes2.default.node,
    string = _propTypes2.default.string;

/**
 *
 * {"categories": ["utility"], "madeFrom": ["LoadingIndicator", "OverlayWrapper"]}
 *
 * A wrapper which can display either a `LoadingIndicator` or `OverlayWrapper`.
 *
 */

var EmptyStateWrapper = (0, _componentTypes.createClass)({
	_isPrivate: true,

	displayName: 'EmptyStateWrapper',

	propTypes: {
		/**
   * Class names that are appended to the defaults.
   */
		className: string,
		/**
   * Any valid React children.
   */
		children: node,
		/**
   * Controls the visibility of the `EmptyMessage`
   */
		isEmpty: bool,
		/**
   * Controls the visibility of the `LoadingMessage`.
   */
		isLoading: bool,
		/**
   * *Child Element*
   *
   * The element to display in the body of the overlay.
   */
		Body: any,
		/**
   * *Child Element*
   *
   * The element to display in the title of the overlay.
   */
		Title: any
	},

	components: {
		/**
   * Body content for the message to display when there is no data.
   */
		Body: (0, _componentTypes.createClass)({
			displayName: 'EmptyStateWrapper.Body',
			propName: 'Body'
		}),
		/**
   * Title text for the message to display when there is no data.
   */
		Title: (0, _componentTypes.createClass)({
			displayName: 'EmptyStateWrapper.Title',
			propName: 'Title'
		})
	},

	render: function render() {
		var _props = this.props,
		    children = _props.children,
		    className = _props.className,
		    isEmpty = _props.isEmpty,
		    isLoading = _props.isLoading,
		    passThroughs = _objectWithoutProperties(_props, ['children', 'className', 'isEmpty', 'isLoading']);

		var emptyMessageBodyProp = (0, _get3.default)((0, _componentTypes.getFirst)(this.props, EmptyStateWrapper.Body), 'props');
		var emptyMessageTitleProp = (0, _get3.default)((0, _componentTypes.getFirst)(this.props, EmptyStateWrapper.Title), 'props', { children: 'You have no data.' });

		return isLoading ? _react2.default.createElement(
			_LoadingIndicator2.default,
			_extends({
				className: cx('&', className),
				isLoading: true
			}, (0, _componentTypes.omitProps)(passThroughs, EmptyStateWrapper, [], false)),
			children
		) : _react2.default.createElement(
			_OverlayWrapper2.default,
			_extends({
				className: cx('&', className),
				hasOverlay: false,
				isVisible: isEmpty
			}, (0, _componentTypes.omitProps)(passThroughs, EmptyStateWrapper, [], false)),
			_react2.default.createElement(
				_OverlayWrapper2.default.Message,
				{ className: cx('&-message-container') },
				_react2.default.createElement('div', { className: cx('&-message-header') }),
				_react2.default.createElement(
					'div',
					{ className: cx('&-message-contents') },
					_react2.default.createElement('header', _extends({}, emptyMessageTitleProp, {
						className: cx('&-message-title', emptyMessageTitleProp.className)
					})),
					emptyMessageBodyProp && _react2.default.createElement('div', emptyMessageBodyProp)
				)
			),
			children
		);
	}
});

exports.default = EmptyStateWrapper;
