// lucid/dist/components/Grid/Grid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Grid');

var string = _propTypes2.default.string,
    bool = _propTypes2.default.bool,
    node = _propTypes2.default.node;

/**
 * {"categories": ["layout"]}
 *
 * This component is designed to be used in Composites as a layout tool.
 * You can use the Grid components themselves or create your own components using the Grid styles from Grid.less.
 * Please see examples for more information.
 */

var Grid = (0, _componentTypes.createClass)({
	displayName: 'Grid',

	components: {
		/**
   * Renders an `<article>` as the grid cell
   */
		Cell: (0, _componentTypes.createClass)({
			displayName: 'Grid.Cell',
			propTypes: {
				/**
     * fill all twelve columns of the primary grid axis
     */
				isFull: bool,
				/**
     * fill six columns of the primary grid axis
     */
				isHalf: bool,
				/**
     * fill four columns of the primary grid axis
     */
				isThird: bool,
				/**
     * fill three columns of the primary grid axis
     */
				isQuarter: bool,

				/**
     * fill 2 columns of 12
     */
				is2: bool,
				/**
     *  fill 3 columns of 12
     */
				is3: bool,
				/**
     * fill 4 columns of 12
     */
				is4: bool,
				/**
     * fill 5 columns of 12
     */
				is5: bool,
				/**
     * fill 6 columns of 12
     */
				is6: bool,
				/**
     * fill 7 columns of 12
     */
				is7: bool,
				/**
     * fill 8 columns of 12
     */
				is8: bool,
				/**
     * fill 9 columns of 12
     */
				is9: bool,
				/**
     * fill 10 columns of 12
     */
				is10: bool,
				/**
     * fill 11 columns of 12
     */
				is11: bool,
				/**
     * offset a grid cell by three columns
     */
				isOffsetQuarter: bool,
				/**
     * offset a grid cell by four columns
     */
				isOffsetThird: bool,
				/**
     * offset a grid cell by six columns
     */
				isOffsetHalf: bool
			}
		})
	},

	propTypes: {
		/**
   * Appended to the component-specific class names set on the root
   * element.
   */
		className: string,

		/**
   * explicitly set the primary axis of the grid to Y
   */
		isVertical: bool,

		/**
   * explicitly set the primary axis of the grid to X
   */
		isHorizontal: bool,

		/**
   * a grid without padding separating grid cells
   */
		isGutterless: bool,

		/**
   * Allow Grids to wrap multiple lines
   */
		isMultiline: bool,

		/**
   * Any valid React component
   */
		children: node
	},

	render: function render() {
		var _props = this.props,
		    className = _props.className,
		    children = _props.children,
		    isVertical = _props.isVertical,
		    isHorizontal = _props.isHorizontal,
		    isGutterless = _props.isGutterless,
		    isMultiline = _props.isMultiline,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'children', 'isVertical', 'isHorizontal', 'isGutterless', 'isMultiline']);

		var cellChildProps = (0, _map3.default)((0, _componentTypes.findTypes)(this.props, Grid.Cell), 'props');

		return _react2.default.createElement(
			'section',
			_extends({}, (0, _componentTypes.omitProps)(passThroughs, Grid), {
				className: cx('&', {
					'&-is-vertical': isVertical,
					'&-is-horizontal': isHorizontal,
					'&-is-gutterless': isGutterless,
					'&-is-multiline': isMultiline
				}, className)
			}),
			(0, _map3.default)(cellChildProps, function (cellChildProp, index) {
				return _react2.default.createElement(
					'article',
					_extends({}, (0, _componentTypes.omitProps)(cellChildProp, Grid.Cell), {
						key: index,
						className: cx('&-Cell', {
							'&-Cell-is-full': cellChildProp.isFull,
							'&-Cell-is-half': cellChildProp.isHalf,
							'&-Cell-is-quarter': cellChildProp.isQuarter,
							'&-Cell-is-third': cellChildProp.isThird,
							'&-Cell-is-2': cellChildProp.is2,
							'&-Cell-is-3': cellChildProp.is3,
							'&-Cell-is-4': cellChildProp.is4,
							'&-Cell-is-5': cellChildProp.is5,
							'&-Cell-is-6': cellChildProp.is6,
							'&-Cell-is-7': cellChildProp.is7,
							'&-Cell-is-8': cellChildProp.is8,
							'&-Cell-is-9': cellChildProp.is9,
							'&-Cell-is-10': cellChildProp.is10,
							'&-Cell-is-11': cellChildProp.is11,
							'&-Cell-is-offset-quarter': cellChildProp.isOffsetQuarter,
							'&-Cell-is-offset-third': cellChildProp.isOffsetThird,
							'&-Cell-is-offset-half': cellChildProp.isOffsetHalf
						}, cellChildProp.className)
					}),
					cellChildProp.children
				);
			}),
			children
		);
	}
});

exports.default = Grid;
