// lucid/dist/components/Sidebar/Sidebar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.SidebarDumb = undefined;

var _isNumber2 = require('lodash/isNumber');

var _isNumber3 = _interopRequireDefault(_isNumber2);

var _first2 = require('lodash/first');

var _first3 = _interopRequireDefault(_first2);

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _noop2 = require('lodash/noop');

var _noop3 = _interopRequireDefault(_noop2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

var _stateManagement = require('../../util/state-management');

var _Sidebar = require('./Sidebar.reducers');

var reducers = _interopRequireWildcard(_Sidebar);

var _SplitVertical = require('../SplitVertical/SplitVertical');

var _SplitVertical2 = _interopRequireDefault(_SplitVertical);

var _ChevronIcon = require('../Icon/ChevronIcon/ChevronIcon');

var _ChevronIcon2 = _interopRequireDefault(_ChevronIcon);

var _GripperVerticalIcon = require('../Icon/GripperVerticalIcon/GripperVerticalIcon');

var _GripperVerticalIcon2 = _interopRequireDefault(_GripperVerticalIcon);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Sidebar');

var any = _propTypes2.default.any,
    bool = _propTypes2.default.bool,
    func = _propTypes2.default.func,
    node = _propTypes2.default.node,
    number = _propTypes2.default.number,
    string = _propTypes2.default.string,
    object = _propTypes2.default.object,
    oneOf = _propTypes2.default.oneOf,
    oneOfType = _propTypes2.default.oneOfType;

/**
 * {"categories": ["layout"]}
 *
 * `Sidebar` renders a collapsible, resizeable side bar panel next to primary content.
 */

var Sidebar = (0, _componentTypes.createClass)({
	displayName: 'Sidebar',

	reducers: reducers,

	propTypes: {
		/**
   * Style object that gets applied to the outer element.
   */
		style: object,

		/**
   * Appended to the component-specific class names set on the root
   * element. Value is run through the `classnames` library.
   */
		className: string,
		/**
   * Direct children must be types {Sidebar.Primary, Sidebar.Bar, Sidebar.Title}.
   * All content is composed as children of these respective elements.
   */
		children: node,
		/**
   * Sets the starting width of the Bar.
   */
		width: oneOfType([number, string]),
		/**
   * Force the Sidebar to be expanded or collapsed.
   */
		isExpanded: bool,
		/**
   * Allows animated expand and collapse behavior.
   */
		isAnimated: bool,
		/**
   * Render the Sidebar to the left or right of primary content.
   */
		position: oneOf(['left', 'right']),
		/**
   * Disable user resizing of the Sidebar.
   */
		isResizeDisabled: bool,
		/**
   * Set the title of the Sidebar. (alias for `Title` and `Sidebar.Title`)
   */
		title: any,
		/**
   * Set the title of the Sidebar. (alias for `title` and `Sidebar.Title`)
   */
		Title: any,
		/**
   * Called when the user is currently resizing the Sidebar.
   *
   * Signature: `(width, { event, props }) => {}`
   */
		onResizing: func,
		/**
   * Called when the user resizes the Sidebar.
   *
   * Signature: `(width, { event, props }) => {}`
   */
		onResize: func,
		/**
   * Called when the user expands or collapses the Sidebar.
   *
   * Signature: `({ event, props }) => {}`
   */
		onToggle: func
	},

	components: {
		/**
   * Content to be placed alongside the Primary pane.
   */
		Bar: (0, _componentTypes.createClass)({
			displayName: 'Sidebar.Bar',
			propTypes: {
				/**
     * Sidebar content. Also can define <Sidebar.Title> here as well.
     */
				children: node,
				/**
     * Set the title of the Sidebar. (alias for `Title` and `Sidebar.Title`)
     */
				title: any,
				/**
     * Set the title of the Sidebar. (alias for `title` and `Sidebar.Title`)
     */
				Title: any,
				/**
     * Adds default padding to the sidebar content.
     */
				hasGutters: bool
			},
			getDefaultProps: function getDefaultProps() {
				return {
					hasGutters: true
				};
			}
		}),

		/**
   * Main pane content that will have a paired `Bar`.
   */
		Primary: (0, _componentTypes.createClass)({
			displayName: 'SplitVertical.Primary',
			propTypes: {
				/**
     * Primary content rendered beside the Sidebar.
     */
				children: node
			}
		}),

		/**
   * Content that will be displayed as the title of the Bar. It's only shown
   * when the user has the Bar expanded.
   */
		Title: (0, _componentTypes.createClass)({
			displayName: 'Sidebar.Title',
			propName: ['Title', 'title'],
			propTypes: {
				/**
     * Sidebar title.
     */
				children: node
			}
		})
	},

	getDefaultProps: function getDefaultProps() {
		return {
			isExpanded: true,
			isAnimated: true,
			width: 250,
			position: 'left',
			isResizeDisabled: false,
			onResizing: _noop3.default,
			onResize: _noop3.default,
			onToggle: _noop3.default
		};
	},
	handleExpanderClick: function handleExpanderClick(event) {
		var onToggle = this.props.onToggle;


		onToggle({ props: this.props, event: event });
	},
	handleResizing: function handleResizing(width, _ref) {
		var event = _ref.event;
		var onResizing = this.props.onResizing;


		onResizing(width, { props: this.props, event: event });
	},
	handleResize: function handleResize(width, _ref2) {
		var event = _ref2.event;
		var onResize = this.props.onResize;


		onResize(width, { props: this.props, event: event });
	},
	render: function render() {
		var _props = this.props,
		    children = _props.children,
		    style = _props.style,
		    className = _props.className,
		    isExpanded = _props.isExpanded,
		    isAnimated = _props.isAnimated,
		    position = _props.position,
		    isResizeDisabled = _props.isResizeDisabled,
		    width = _props.width,
		    passThroughs = _objectWithoutProperties(_props, ['children', 'style', 'className', 'isExpanded', 'isAnimated', 'position', 'isResizeDisabled', 'width']);

		var primaryProps = (0, _get3.default)((0, _first3.default)((0, _componentTypes.filterTypes)(children, Sidebar.Primary)), 'props', {}); // props from first Primary
		var barProps = (0, _get3.default)((0, _first3.default)((0, _componentTypes.filterTypes)(children, Sidebar.Bar)), 'props', {}); // props from first Bar
		var titleProps = (0, _get3.default)((0, _componentTypes.findTypes)(barProps, Sidebar.Title).concat((0, _componentTypes.findTypes)(this.props, Sidebar.Title)), // get titles from Bar and parent Sidebar
		'[0].props', // select props from the first title element
		_react2.default.createElement(
			Sidebar.Title,
			null,
			'Title'
		).props // default props
		);

		var PrimaryPane = void 0,
		    BarPane = void 0; // using Left/Right Pane as primary depends on position
		if (position !== 'right') {
			PrimaryPane = _SplitVertical2.default.RightPane;
			BarPane = _SplitVertical2.default.LeftPane;
		} else {
			PrimaryPane = _SplitVertical2.default.LeftPane;
			BarPane = _SplitVertical2.default.RightPane;
		}

		return _react2.default.createElement(
			_SplitVertical2.default,
			_extends({}, (0, _componentTypes.omitProps)(passThroughs, Sidebar, [], false), {
				style: _extends({
					minWidth: isExpanded ? (0, _isNumber3.default)(width) ? width + 6 : 'calc(' + width + ' + 6px)' : null
				}, style),
				className: cx('&', {
					'&-is-resize-disabled': isResizeDisabled,
					'&-is-position-right': position === 'right',
					'&-is-position-left': position !== 'right'
				}, className),
				isAnimated: isAnimated,
				isExpanded: isExpanded,
				collapseShift: 33 // leave 33px of sidebar to stick out when collapsed
				, onResizing: this.handleResizing,
				onResize: this.handleResize
			}),
			_react2.default.createElement(
				BarPane,
				_extends({}, (0, _componentTypes.omitProps)(barProps, Sidebar.Bar, [], false), {
					className: cx('&-Bar', barProps.className),
					width: width,
					style: {
						overflow: isExpanded ? 'auto' : 'hidden'
					}
				}),
				_react2.default.createElement('div', { className: cx('&-Bar-overlay') }),
				_react2.default.createElement(
					'div',
					{ className: cx('&-Bar-header') },
					_react2.default.createElement('div', _extends({}, titleProps, {
						className: cx('&-Bar-Title', titleProps.className)
					})),
					_react2.default.createElement(
						'div',
						{
							className: cx('&-expander'),
							onMouseDown: this.handleExpanderClick
						},
						_react2.default.createElement(_ChevronIcon2.default, {
							direction: isExpanded && position === 'right' || !isExpanded && position !== 'right' ? 'right' : 'left'
						})
					)
				),
				_react2.default.createElement(
					'div',
					{
						className: cx('&-Bar-content', {
							'&-Bar-content-has-gutters': barProps.hasGutters
						})
					},
					barProps.children
				)
			),
			_react2.default.createElement(
				_SplitVertical2.default.Divider,
				{ className: cx('&-Divider'), style: { width: 7 } },
				_react2.default.createElement(_GripperVerticalIcon2.default, { className: cx('&-Divider-gripper') })
			),
			_react2.default.createElement(PrimaryPane, _extends({}, primaryProps, {
				className: cx('&-Primary', primaryProps.className),
				isPrimary: true
			}))
		);
	}
});

exports.default = (0, _stateManagement.buildHybridComponent)(Sidebar);
exports.SidebarDumb = Sidebar;
