// lucid/dist/components/Paginator/Paginator.reducers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.TextField = exports.SingleSelect = undefined;

var _clamp2 = require('lodash/clamp');

var _clamp3 = _interopRequireDefault(_clamp2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.onPageSelect = onPageSelect;
exports.onPageSizeSelect = onPageSizeSelect;

var _SingleSelect = require('../SingleSelect/SingleSelect.reducers');

var SingleSelect = _interopRequireWildcard(_SingleSelect);

var _TextField = require('../TextField/TextField.reducers');

var TextField = _interopRequireWildcard(_TextField);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function onPageSelect(state, pageIndex, totalPages) {
	return _extends({}, state, {
		selectedPageIndex: (0, _clamp3.default)(pageIndex, 0, totalPages - 1)
	});
}

function onPageSizeSelect(state, selectedPageSizeIndex) {
	return _extends({}, state, {
		selectedPageIndex: 0,
		selectedPageSizeIndex: selectedPageSizeIndex,
		SingleSelect: SingleSelect.onSelect(state.SingleSelect, selectedPageSizeIndex)
	});
}

exports.SingleSelect = SingleSelect;
exports.TextField = TextField;
