// lucid/dist/components/Bar/Bar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _startsWith2 = require('lodash/startsWith');

var _startsWith3 = _interopRequireDefault(_startsWith2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

var _charts = require('../../constants/charts');

var chartConstants = _interopRequireWildcard(_charts);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Bar');

var number = _propTypes2.default.number,
    bool = _propTypes2.default.bool,
    string = _propTypes2.default.string,
    object = _propTypes2.default.object;

/**
 * {"categories": ["visualizations", "geoms"]}
 *
 * *For use within an `svg`*
 *
 * Bars are typically used for bar charts and are pretty much a thin wrapper
 * around svg rects.
 *
 */

var Bar = (0, _componentTypes.createClass)({
	displayName: 'Bar',

	propTypes: {
		/**
   * Passed through to the root element.
   */
		style: object,
		/**
   * Appended to the component-specific class names set on the root element.
   */
		className: string,
		/**
   * x coordinate.
   */
		x: number,
		/**
   * y coordinate.
   */
		y: number,
		/**
   * Height of the bar.
   */
		height: number,
		/**
   * Width of the bar.
   */
		width: number,
		/**
   * Determines if the bar has a white stroke around it.
   */
		hasStroke: bool,
		/**
   * Strings should match an existing color class unless they start with a
   * '#' for specific colors. E.g.:
   *
   * - `COLOR_0`
   * - `COLOR_GOOD`
   * - `'#123abc'`
   */
		color: string
	},

	getDefaultProps: function getDefaultProps() {
		return {
			x: 0,
			y: 0,
			height: 0,
			width: 0,
			color: chartConstants.COLOR_0
		};
	},
	render: function render() {
		var _props = this.props,
		    className = _props.className,
		    color = _props.color,
		    hasStroke = _props.hasStroke,
		    height = _props.height,
		    width = _props.width,
		    style = _props.style,
		    x = _props.x,
		    y = _props.y,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'color', 'hasStroke', 'height', 'width', 'style', 'x', 'y']);

		var isCustomColor = (0, _startsWith3.default)(color, '#');
		var colorStyle = isCustomColor ? { fill: color } : null;

		return _react2.default.createElement('rect', _extends({}, (0, _componentTypes.omitProps)(passThroughs, Bar), {
			className: cx(className, '&', _defineProperty({
				'&-has-stroke': hasStroke
			}, '&-' + color, !isCustomColor)),
			x: x,
			y: y,
			height: height,
			width: width,
			style: _extends({}, style, colorStyle)
		}));
	}
});

exports.default = Bar;
